# 쓰레기통 데이터 csv 파일 불러오기


In [47]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
ls

drive/  sample_data/


In [48]:
import pandas as pd

# 마이크로소프트는 CP949 인코딩 방식을 사용하기 때문에 UTF-8 인코딩을 적용할 수 없다(UTF-8은 Mac, Linux)
df = pd.read_csv('drive/MyDrive/Colab Notebooks/trashcan/trashcan_data_15.csv', encoding='utf-8')

In [28]:
df

연번 자치구명   도로명                            상세주소     설치지점   종류  정류장번호  \
0        1  종로구   사직로                       경복궁역 4번출구  지하철역 입구   일반    NaN   
1        2  종로구   사직로                       경복궁역 4번출구  지하철역 입구  재활용    NaN   
2        3  종로구  자하문로                         자하문로 28  도로(가로)변   일반    NaN   
3        4  종로구  자하문로                         자하문로 28  도로(가로)변  재활용    NaN   
4        5  종로구  자하문로                         자하문로 44  도로(가로)변   일반    NaN   
...    ...  ...   ...                             ...      ...  ...    ...   
4894  5371  도봉구   해등로            해등로3길 41 삼환빌라 앞 횡단보도  도로(가로)변   일반    NaN   
4895  5372  도봉구   해등로          해등로 103  창원초등학교 앞 횡단보도  도로(가로)변   일반    NaN   
4896  5373  도봉구  도봉산길         도봉산길 27 도봉고등학교 건너편 횡단보도  도로(가로)변   일반    NaN   
4897  5374  도봉구   도봉로  도봉동 620-23 신도봉사거리 도봉중학교방향 횡단보도  도로(가로)변   일반    NaN   
4898  5375  도봉구   해등로        해등로 32 창1동 서울가든아파트 버스정류장    버스정류장   일반    NaN   

             위도          경도  
0           NaN         NaN  
1           NaN         NaN  
2     37.578567  126.971970  
3     37.578567  126.971970  
4     37.579972  126.971550  
...         ...         ...  
4894  37.645712  127.042082  
4895  37.652644  127.039980  
4896  37.686755  127.042578  
4897  37.670980  127.043439  
4898  37.501074  126.860358  

[4899 rows x 9 columns]

# 버스 정류장 데이터

In [ ]:

# 마이크로소프트는 CP949 인코딩 방식을 사용하기 때문에 UTF-8 인코딩을 적용할 수 없다(UTF-8은 Mac, Linux)
bus_stop_data = pd.read_csv('drive/MyDrive/Colab Notebooks/trashcan/bus_stop_data.csv', encoding='CP949')
bus_stop_data

노드 ID  정류소번호           정류소명         X좌표        Y좌표 정류소 타입
0      100000001   1001        종로2가사거리  126.987752  37.569808   중앙차로
1      100000002   1002    창경궁.서울대학교병원  126.996522  37.579433   중앙차로
2      100000003   1003      명륜3가.성대입구  126.998251  37.582581   중앙차로
3      100000004   1004       종로2가.삼일교  126.987613  37.568579   중앙차로
4      100000005   1005  혜화동로터리.여운형활동터  127.001744  37.586243   중앙차로
...          ...    ...            ...         ...        ...    ...
11285  124000334  25995          우성아파트  127.139339  37.550386   일반차로
11286  124000333  25996          우성아파트  127.140046  37.550643   일반차로
11287  124000332  25997           조일약국  127.123596  37.533630   일반차로
11288  124000331  25998           성내시장  127.125497  37.536155   일반차로
11289  124000330  25999    천호우체국.로데오거리  127.127337  37.540343   일반차로

[11290 rows x 6 columns]

# 중복된 데이터 제거하기

In [ ]:
# 자치구명, 도로명, 상세주소, 설치지점, 종류가 일치하는 중복된 행을 제거합니다.
df = df.drop_duplicates(subset=['자치구명', '도로명', '상세주소', '설치지점', '종류'])


# 택시정류장 골라내기
- 상세주소에 "택시"라는 말이 들어갈 경우 설치 지점을 "택시정류장"으로 변경


In [ ]:
for i in range(5375):
  if "택시" in df['상세주소'][i] :
    df['설치지점'][i] = "택시정류장"


# 설치지점이 상가, 상호명으로 되어 있을 경우 상세주소에 붙이기

- 인덱스 번호가 5095 ~ 5374번 까지 확인(연번은 5096 ~ 5375)
- 설치지점은 "도로(가로)변"으로 변경

In [ ]:
for i in range(5095, 5375) :
  df['상세주소'][i] = df['상세주소'][i]+" "+df['설치지점'][i]
  df['설치지점'][i] = "도로(가로)변"
  if(df['상세주소'][i].strip()[-1] == "앞") :
     df['상세주소'][i] = df['상세주소'][i][:-1]

In [ ]:
# 공백 제거
for i in range(5375) :
  df['상세주소'][i] = df['상세주소'][i].strip()


# 버스정류장으로 설치지점 바꾸기 (설치지점)

- 설치 지점에 "버스" 라는 말이 들어갈 경우 설치 지점을 "버스정류장으로 변경

In [ ]:
for i in range(5375):
  if "버스" in df['설치지점'][i] :
    df['설치지점'][i] = "버스정류장"
    print(df.loc[i])


# 버스정류장으로 설치지점 바꾸기 (상세주소)

- 상세 주소에 "버스" 라는 말이 들어갈 경우 설치 지점을 "버스정류장으로 변경


In [ ]:
for i in range(5375):
  if "버스" in df['상세주소'][i] :
    df['설치지점'][i] = "버스정류장"
    print(df.loc[i])


# 설치지점이 "버스정류장"인 데이터에서 정류장번호 알아내기
- 상세주소에서 @@@@@ 또는 @@-@@@ 형식을 가진 데이터를 정류장 번호
- 앞 뒤로 숫자가 더 있는 경우 제외
- 첫번째 글자가 0, 1, 2 가 아닌 경우 제외





In [ ]:
import re # 정규표현 처리 라이브러리


# 정규 표현식 패턴 정의 @@@@@ 또는 @@-@@@
pattern = r'(\d{5})|(\d{2}-\d{3})'


def extract_bus_stop_number(row):
    address = row['상세주소']
    if isinstance(address, str):  # 데이터가 문자열인지 확인
        match = re.search(pattern, address)  # address에서 정규표현식 pattern 과 일치하는 부분이 있다면 그 부분을 추출
        if match:
            start_index = match.start()
            if start_index != 0 and address[start_index-1].isdigit() :
              print("상세 주소에 해당 패턴 없음 : ", address)
              return None

            last_index = match.end()
            if last_index != len(address) and address[last_index].isdigit() :
              print("상세 주소에 해당 패턴 없음 : ", address)
              return None
            if match.group()[0] not in ['0', '1', '2']:
              print(row)
              print("정류소 번호가 아님 : ", match.group())
              return None

            bus_stop_number = match.group()  # 패턴에서 일치하여 추출한 문자열 반환
            # '-' 기호가 있는 경우 제거하고 '@@@@@' 형식으로 변환
            bus_stop_number = bus_stop_number.replace('-', '')
            return '{:05}'.format(int(bus_stop_number)) # 앞부분이 0으로 채워지도록 포맷팅
    return None  # 일치하는 패턴이 없거나 데이터가 문자열이 아닌 경우 None을 반환

# '정류장번호' 열에 함수를 적용하여 정류장번호 알아내기
df['정류장번호'] = df.apply(extract_bus_stop_number, axis=1) # 기존의 Column 값을 연산하여 새 Column을 추가하는 경
# axis =0 은 row, axis = 1은 column

df


# 버스 정류장 번호를 이용해서 위도, 경도 알아내기
- 서울시 버스정류소 위치정보 csv
- https://data.seoul.go.kr/dataList/OA-15067/S/1/datasetView.do#


In [ ]:

# 정류장번호를 기반으로 정류소 정보를 찾아서 경도와 위도를 추가합니다.
for index, row in df.iterrows():
    정류장번호 = row['정류장번호']
    # bus_stop_data.csv에서 정류장번호에 해당하는 행을 찾습니다.
    bus_stop_info = bus_stop_data[bus_stop_data['정류소번호'] == 정류장번호]
    # 정류소 정보가 있는 경우 해당 정보의 경도(X좌표)와 위도(Y좌표)를 가져와 업데이트합니다.
    if not bus_stop_info.empty:
        df.at[index, '경도'] = bus_stop_info.iloc[0]['X좌표']
        df.at[index, '위도'] = bus_stop_info.iloc[0]['Y좌표']

df

# 업데이트된 데이터를 새로운 CSV 파일로 저장
# df.to_csv('updated_trashcan_data.csv', index=False, encoding='utf-8')

# print("업데이트된 데이터가 updated_trashcan_data.csv 파일로 저장되었습니다.")

연번 자치구명   도로명                            상세주소     설치지점   종류  정류장번호  \
0        1  종로구   사직로                       경복궁역 4번출구  지하철역 입구   일반    NaN   
1        2  종로구   사직로                       경복궁역 4번출구  지하철역 입구  재활용    NaN   
2        3  종로구  자하문로                         자하문로 28  도로(가로)변   일반    NaN   
3        4  종로구  자하문로                         자하문로 28  도로(가로)변  재활용    NaN   
4        5  종로구  자하문로                         자하문로 44  도로(가로)변   일반    NaN   
...    ...  ...   ...                             ...      ...  ...    ...   
4894  5371  도봉구   해등로            해등로3길 41 삼환빌라 앞 횡단보도  도로(가로)변   일반    NaN   
4895  5372  도봉구   해등로          해등로 103  창원초등학교 앞 횡단보도  도로(가로)변   일반    NaN   
4896  5373  도봉구  도봉산길         도봉산길 27 도봉고등학교 건너편 횡단보도  도로(가로)변   일반    NaN   
4897  5374  도봉구   도봉로  도봉동 620-23 신도봉사거리 도봉중학교방향 횡단보도  도로(가로)변   일반    NaN   
4898  5375  도봉구   해등로        해등로 32 창1동 서울가든아파트 버스정류장    버스정류장   일반    NaN   

      위도  경도  
0    NaN NaN  
1    NaN NaN  
2    NaN NaN  
3    NaN NaN  
4    NaN NaN  
...   ..  ..  
4894 NaN NaN  
4895 NaN NaN  
4896 NaN NaN  
4897 NaN NaN  
4898 NaN NaN  

[4899 rows x 9 columns]

# (확인용)정류장 번호로 위도, 경도 찾아내지 못한 데이터 개수

In [ ]:
# 정류장 번호는 있지만 위도와 경도가 NaN인 데이터의 개수 구하기
no_coordinates_count = len(df[(~df['정류장번호'].isnull()) & (df['위도'].isnull() | df['경도'].isnull())])

print(f"정류장 번호는 있지만 위도와 경도가 없는 데이터의 총 개수: {no_coordinates_count}개")


정류장 번호는 있지만 위도와 경도가 없는 데이터의 총 개수: 12개


In [ ]:
# 정류장 번호는 있지만 위도와 경도가 NaN인 데이터 출력
missing_coordinates_data = df[(~df['정류장번호'].isnull()) & (df['위도'].isnull() | df['경도'].isnull())]

print("정류장 번호는 있지만 위도와 경도가 없는 데이터:")
missing_coordinates_data

정류장 번호는 있지만 위도와 경도가 없는 데이터:


연번  자치구명   도로명                                    상세주소   설치지점  \
450    477   용산구  한강대로                     남영역 마을버스정류장 (03130)  버스정류장   
695    754   성동구   마조로                               행당동19-100  버스정류장   
2095  2281  서대문구   신촌로                         신촌로 아현역 (22062)  버스정류장   
2097  2283  서대문구   신촌로                        신촌로 웨딩타운 (22061)  버스정류장   
2574  2794    강서   강서로                        그랜드마트 정류장앞16-337  버스정류장   
2575  2795    강서   강서로                        그랜드마트 정류장앞16-337  버스정류장   
3110  3392  영등포구  영등포동             경방타임스퀘어,신세계백화점버스정류장(19-110)  버스정류장   
3111  3393  영등포구  영등포동                     신세계백화점버스정류장(19-111)  버스정류장   
3138  3420  영등포구   당산동                     선유고등학교버스정류장(28-017)  버스정류장   
4406  4836   송파구  위례대로                   스타필드시티.위례스타힐스(28-533)  버스정류장   
4407  4837   송파구  위례대로                   스타필드시티.위례스타힐스(28-533)  버스정류장   
4450  4907   강동구   천중로  길동신동아아파트(길동신동아아파트사거리 방면) 버스정류장(25-502)  버스정류장   

             종류    정류장번호  위도  경도  
450    일반 03130   3130.0 NaN NaN  
695          일반  19100.0 NaN NaN  
2095         일반  22062.0 NaN NaN  
2097         일반  22061.0 NaN NaN  
2574         일반  16337.0 NaN NaN  
2575       재활용   16337.0 NaN NaN  
3110         일반  19110.0 NaN NaN  
3111         일반  19111.0 NaN NaN  
3138         일반  28017.0 NaN NaN  
4406         일반  28533.0 NaN NaN  
4407        재활용  28533.0 NaN NaN  
4450         일반  25502.0 NaN NaN

# (확인용)버스정류장인데 버스정류장 번호가 없는 데이터 개수


In [ ]:
# 설치지점이 "버스정류장"이고 정류장번호가 없는 데이터 찾기
missing_bus_stop_data = df[(df['설치지점'] == '버스정류장') & (df['정류장번호'].isnull())]

# 개수 출력
print("정류장번호가 없는 버스정류장 데이터의 개수:", len(missing_bus_stop_data))

# 해당 데이터 출력
print("정류장번호가 없는 버스정류장 데이터:")
missing_bus_stop_data


정류장번호가 없는 버스정류장 데이터의 개수: 1202
정류장번호가 없는 버스정류장 데이터:


연번 자치구명   도로명                              상세주소   설치지점   종류  정류장번호  \
6        7  종로구  자하문로                  자하문로 68(효자동 정류소)  버스정류장   일반    NaN   
7        8  종로구  자하문로                  자하문로 68(효자동 정류소)  버스정류장  재활용    NaN   
43      46  종로구    종로                        종로1가 버스정류장  버스정류장   일반    NaN   
44      47  종로구    종로                        종로1가 버스정류장  버스정류장  재활용    NaN   
50      54  종로구  삼일대로                        삼일대로 버스정류장  버스정류장   일반    NaN   
...    ...  ...   ...                               ...    ...  ...    ...   
4677  5151  도봉구   도봉로            도봉로 859 건너편 도봉삼환 버스정류장  버스정류장   일반    NaN   
4829  5305  도봉구   노해로  노해로41가길 16 e편한세상 102동 앞 둘리 버스정류장  버스정류장   일반    NaN   
4890  5367  도봉구   해등로            해등로 139 미소애아파트 앞 버스정류장  버스정류장   일반    NaN   
4891  5368  도봉구   해등로            창동 347-5 창1동주민센터 버스정류장  버스정류장   일반    NaN   
4898  5375  도봉구   해등로          해등로 32 창1동 서울가든아파트 버스정류장  버스정류장   일반    NaN   

      위도  경도  
6    NaN NaN  
7    NaN NaN  
43   NaN NaN  
44   NaN NaN  
50   NaN NaN  
...   ..  ..  
4677 NaN NaN  
4829 NaN NaN  
4890 NaN NaN  
4891 NaN NaN  
4898 NaN NaN  

[1202 rows x 9 columns]

# 정류장 번호가 없는 버스정류장 위도, 경도 알아내기(1)
- 상세주소와 정류소명이 같은 데이터를 찾아내서 위도, 경도 추출




In [ ]:
update_count = 0  # 업데이트된 데이터의 개수를 저장할 변수

# 정류장 번호가 존재하지 않는 데이터의 상세주소를 기반으로 정류소명과 일치하는 데이터를 찾아 X좌표를 경도, Y좌표를 위도에 입력
for index, row in df.iterrows():
    정류장번호 = row['정류장번호']
    상세주소 = row['상세주소']

    # 정류장 번호가 존재하지 않는 경우
    if pd.isnull(정류장번호):
        # 상세주소를 기반으로 정류소명과 일치하는 데이터 찾기
        matching_bus_stop = bus_stop_data[bus_stop_data['정류소명'] == 상세주소]
        # 일치하는 데이터가 있을 경우에만 경도와 위도 업데이트
        if not matching_bus_stop.empty:
            df.at[index, '경도'] = matching_bus_stop.iloc[0]['X좌표']
            df.at[index, '위도'] = matching_bus_stop.iloc[0]['Y좌표']
            update_count += 1  # 업데이트된 데이터의 개수를 증가

print(f"총 {update_count}개의 데이터가 업데이트되었습니다.")

총 24개의 데이터가 업데이트되었습니다.


# 정류장 번호가 없는 버스정류장 위도, 경도 알아내기(2)
- 상세주소에서 버스정류장 앞에 존재하는 단어와 정류소명과 동일한 데이터를 찾아내 위도, 경도 추출





In [ ]:
# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# trashcan_data에서 설치지점이 "버스정류장"이고 위도, 경도가 존재하지 않는 데이터 처리
for index, row in df.iterrows():
    설치지점 = row['설치지점']
    위도 = row['위도']
    경도 = row['경도']
    상세주소 = row['상세주소']

    # 설치지점이 "버스정류장"이고 위도, 경도가 존재하지 않는 경우
    if 설치지점 == '버스정류장' and pd.isnull(위도) and pd.isnull(경도):
        # 상세주소에서 "버스정류장"이라는 단어가 포함되어 있는 경우 처리
        address_parts = 상세주소.split()
        if "버스정류장" in address_parts:
            # 띄어쓰기로 split하여 "버스정류장" 문자열의 인덱스 -1 앞의 인덱스 문자열 추출
            bus_stop_index = address_parts.index("버스정류장")
            bus_stop_name = address_parts[bus_stop_index-1]
            if bus_stop_name and bus_stop_name[0] == '(':
               bus_stop_name = bus_stop_name[1:]  # 첫 번째 문자 '(' 제거
            # bus_stop_data에서 정류소명과 일치하는 데이터 찾기
            matching_bus_stop = bus_stop_data[bus_stop_data['정류소명'] == bus_stop_name]
            # 일치하는 데이터가 있을 경우에만 경도와 위도 업데이트
            if not matching_bus_stop.empty:
                print(bus_stop_name)
                df.at[index, '경도'] = matching_bus_stop.iloc[0]['X좌표']
                df.at[index, '위도'] = matching_bus_stop.iloc[0]['Y좌표']
                updated_count += 1  # 업데이트된 데이터의 개수 증가

# 업데이트된 데이터의 개수 출력
print(f"총 {updated_count}개의 데이터가 업데이트되었습니다.")


종로1가
종로1가
구의사거리
구의사거리
삼거리
삼거리
경남아너스빌
대림시장
대림시장
은평연세병원
은평연세병원
와산교
와산교
숭실고입구
숭실고입구
연신내역
연신내역
대일고등학교
대일고등학교
한보구암마을아파트
태영송화아파트
마곡나루역
마곡엠밸리14단지
공항동주민센터
공항동천주교회
신방화사거리
방화동도시개발11단지
생태공원
서광아파트
서남물재생센터
두산아파트
두산아파트
여의도중학교
여의도중학교
일동제약사거리
서래마을입구
무역센터
무역센터
정병원
우이1교
창1동주민센터
서울가든아파트
총 42개의 데이터가 업데이트되었습니다.


# 나머지 데이터 `상세주소`를 카카오 맵 api에 `주소`로 검색
- 위도, 경도가 없다면
- 주소로 위도, 경도 알아내기
- 카카오 맵 api에 검색



In [ ]:
import requests

# Kakao Map API 키
KAKAO_API_KEY = '24a11401d72ce55f0cc05a169bec666e'


# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# 위도와 경도가 없는 데이터를 카카오 맵 API로 검색하여 업데이트
for index, row in df.iterrows():
    위도 = row['위도']
    경도 = row['경도']
    상세주소 = row['상세주소']

    # 위도와 경도가 없는 경우
    if pd.isnull(위도) or pd.isnull(경도):
        # 상세주소가 없으면 건너뜀
        if pd.isnull(상세주소):
            continue

        # Kakao Map API로 주소 검색
        url = f'https://dapi.kakao.com/v2/local/search/address.json?query={상세주소}'
        headers = {'Authorization': f'KakaoAK {KAKAO_API_KEY}'}
        response = requests.get(url, headers=headers)

        # API 응답 확인
        if response.status_code == 200:
            data = response.json()
            if data['documents']:
                # 첫 번째 결과의 위도와 경도를 업데이트
                df.at[index, '위도'] = data['documents'][0]['y']
                df.at[index, '경도'] = data['documents'][0]['x']
                print(f'인덱스 {index}의 데이터의 위도와 경도를 업데이트했습니다.')
                print(f'주소: ', 상세주소)
                print(f'위도: ', df.at[index, '위도'])
                print(f'경도:',  df.at[index, '경도'])
                updated_count += 1
            else:
                print(상세주소 , '의 데이터에 대한 검색 결과가 없습니다.')
        else:
            print(f'인덱스 {index}의 데이터에 대한 API 호출에 실패했습니다.')

# 업데이트된 데이터의 개수 출력
print(f'업데이트된 데이터의 개수: {updated_count}')


스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
주소:  월드컵북로 74 현대자동차 성산대리점 앞
위도:  37.5606073905793
경도: 126.916371238301
인덱스 2260의 데이터의 위도와 경도를 업데이트했습니다.
주소:  월드컵북로 136 하나은행 옆
위도:  37.5650251234037
경도: 126.9124736579
인덱스 2262의 데이터의 위도와 경도를 업데이트했습니다.
주소:  월드컵북로56길 9 옆 가온문화공원
위도:  37.581384890718
경도: 126.887439627184
인덱스 2266의 데이터의 위도와 경도를 업데이트했습니다.
주소:  상암산로1길 52 월드컵파크5단지 508동 앞
위도:  37.5808073691977
경도: 126.88539532769
인덱스 2268의 데이터의 위도와 경도를 업데이트했습니다.
주소:  월드컵로202 마포구청역 1번 출구 앞
위도:  37.5652870365795
경도: 126.902464118514
인덱스 2269의 데이터의 위도와 경도를 업데이트했습니다.
주소:  월드컵로204 마포구 청사 맞은편 횡단보도
위도:  37.5653992058145
경도: 126.9021317644
인덱스 2270의 데이터의 위도와 경도를 업데이트했습니다.
주소:  상암산로1길 26 상암중학교 앞
위도:  37.5790569198061
경도: 126.887550793573
구청 광장 의 데이터에 대한 검색 결과가 없습니다.
구청 광장 4구 재활용 의 데이터에 대한 검색 결과가 없습니다.
인덱스 2273의 데이터의 위도와 경도를 업데이트했습니다.
주소:  월드컵북로381 월드컵파크4단지 401동 앞
위도:  37.5779528826091
경도: 126.889887582062
마포농수산물시장 3거리 홈플러스 입구 의 데이터에 대한 검색 결과가 없습니다.
평화공원 주차장입구 횡단보도 맞은편 의 데이터에 대한 검색 결과가 없습니다.
월드컵경기장 사거리 의 데이터에 대한 검색 결과가 

# 나머지 데이터 `상세주소`를 카카오 맵 api에 `키워드`로 검색
- 위도, 경도가 없다면
- 주소로 위도, 경도 알아내기
- 카카오 맵 api에 검색



In [ ]:
import requests

# Kakao Map API 키
KAKAO_API_KEY = '24a11401d72ce55f0cc05a169bec666e'


# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# 위도와 경도가 없는 데이터를 카카오 맵 API로 검색하여 업데이트
for index, row in df.iterrows():
    위도 = row['위도']
    경도 = row['경도']
    상세주소 = row['상세주소']

    # 위도와 경도가 없는 경우
    if pd.isnull(위도) or pd.isnull(경도):
        # 상세주소가 없으면 건너뜀
        if pd.isnull(상세주소):
            continue

        # Kakao Map API로 주소 검색
        url = f'https://dapi.kakao.com/v2/local/search/keyword.json?query={상세주소}'
        headers = {'Authorization': f'KakaoAK {KAKAO_API_KEY}'}
        response = requests.get(url, headers=headers)

        # API 응답 확인
        if response.status_code == 200:
            data = response.json()
            if data['documents']:
                # 첫 번째 결과의 위도와 경도를 업데이트
                df.at[index, '위도'] = data['documents'][0]['y']
                df.at[index, '경도'] = data['documents'][0]['x']
                print(f'인덱스 {index}의 데이터의 위도와 경도를 업데이트했습니다.')
                print(f'주소: ', 상세주소)
                print(f'위도: ', df.at[index, '위도'])
                print(f'경도:',  df.at[index, '경도'])
                updated_count += 1
            else:
                print(상세주소 , '의 데이터에 대한 검색 결과가 없습니다.')
        else:
            print(f'인덱스 {index}의 데이터에 대한 API 호출에 실패했습니다.')

# 업데이트된 데이터의 개수 출력
print(f'업데이트된 데이터의 개수: {updated_count}')


스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
경도: 127.099249574265
인덱스 2876의 데이터의 위도와 경도를 업데이트했습니다.
주소:  공원로6길
위도:  37.50218730704493
경도: 126.89166152290304
인덱스 2877의 데이터의 위도와 경도를 업데이트했습니다.
주소:  공원로54
위도:  37.50208379700182
경도: 126.89079764415249
신도림역2번(택시) 의 데이터에 대한 검색 결과가 없습니다.
새말로117-6(횡단보도) 의 데이터에 대한 검색 결과가 없습니다.
신도림역2번(신호등) 의 데이터에 대한 검색 결과가 없습니다.
신도림역2번(자전거) 의 데이터에 대한 검색 결과가 없습니다.
신도림역2번(환승) 의 데이터에 대한 검색 결과가 없습니다.
인덱스 2883의 데이터의 위도와 경도를 업데이트했습니다.
주소:  경인로572
위도:  37.5032690729499
경도: 126.881006458503
인덱스 2884의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로중앙로 68
위도:  37.4958289897032
경도: 126.888839103311
인덱스 2885의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로중앙로28길 65
위도:  37.50038177078585
경도: 126.88883687802908
구로중앙로28길 108 의 데이터에 대한 검색 결과가 없습니다.
인덱스 2887의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로동514-2
위도:  37.4985242396083
경도: 126.88635623368
인덱스 2888의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로동3-64
위도:  37.5073638319324
경도: 126.892925294001
인덱스 2889의 데이터의 위도와 경도를 업데이트했습니다.
주소:  공원로27
위도:  37.49933160529627
경도: 126.89120990191064
인덱스 2890의

# 나머지 데이터 `자치구명 +” “ +도로명+” “+ 상세주소`를 카카오 맵 api에 `주소`로 검색
- 위도, 경도가 없다면
- 주소로 위도, 경도 알아내기
- 카카오 맵 api에 검색



In [ ]:
import requests

# Kakao Map API 키
KAKAO_API_KEY = '24a11401d72ce55f0cc05a169bec666e'


# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# 위도와 경도가 없는 데이터를 카카오 맵 API로 검색하여 업데이트
for index, row in df.iterrows():
    위도 = row['위도']
    경도 = row['경도']
    자치구명 = row['자치구명']
    도로명 = row['도로명']
    상세주소 = row['상세주소']

    # 위도와 경도가 없는 경우
    if pd.isnull(위도) or pd.isnull(경도):
        주소 = f'{자치구명} {도로명} {상세주소}'

        # Kakao Map API로 주소 검색
        url = f'https://dapi.kakao.com/v2/local/search/address.json?query={상세주소}'
        headers = {'Authorization': f'KakaoAK {KAKAO_API_KEY}'}
        response = requests.get(url, headers=headers)

        # API 응답 확인
        if response.status_code == 200:
            data = response.json()
            if data['documents']:
                # 첫 번째 결과의 위도와 경도를 업데이트
                df.at[index, '위도'] = data['documents'][0]['y']
                df.at[index, '경도'] = data['documents'][0]['x']
                print(f'인덱스 {index}의 데이터의 위도와 경도를 업데이트했습니다.')
                print(f'주소: ', 주소)
                print(f'위도: ', df.at[index, '위도'])
                print(f'경도:',  df.at[index, '경도'])
                updated_count += 1
            else:
                print(f'{주소}의 데이터에 대한 검색 결과가 없습니다.')
        else:
            print(f'인덱스 {index}의 데이터에 대한 API 호출에 실패했습니다.')

# 업데이트된 데이터의 개수 출력
print(f'업데이트된 데이터의 개수: {updated_count}')


스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
경도: 126.88539532769
인덱스 2268의 데이터의 위도와 경도를 업데이트했습니다.
주소:  마포구 월드컵로 월드컵로202 마포구청역 1번 출구 앞
위도:  37.5652870365795
경도: 126.902464118514
인덱스 2269의 데이터의 위도와 경도를 업데이트했습니다.
주소:  마포구 월드컵로 월드컵로204 마포구 청사 맞은편 횡단보도
위도:  37.5653992058145
경도: 126.9021317644
인덱스 2270의 데이터의 위도와 경도를 업데이트했습니다.
주소:  마포구 상암산로 상암산로1길 26 상암중학교 앞
위도:  37.5790569198061
경도: 126.887550793573
마포구 월드컵로 구청 광장의 데이터에 대한 검색 결과가 없습니다.
마포구 월드컵로 구청 광장 4구 재활용의 데이터에 대한 검색 결과가 없습니다.
인덱스 2273의 데이터의 위도와 경도를 업데이트했습니다.
주소:  마포구 월드컵북로 월드컵북로381 월드컵파크4단지 401동 앞
위도:  37.5779528826091
경도: 126.889887582062
마포구 월드컵로 마포농수산물시장 3거리 홈플러스 입구의 데이터에 대한 검색 결과가 없습니다.
마포구 월드컵로 평화공원 주차장입구 횡단보도 맞은편의 데이터에 대한 검색 결과가 없습니다.
마포구 월드컵로 월드컵경기장 사거리의 데이터에 대한 검색 결과가 없습니다.
마포구 월드컵로 마포구청역 5번 출구 뒤의 데이터에 대한 검색 결과가 없습니다.
인덱스 2286의 데이터의 위도와 경도를 업데이트했습니다.
주소:  마포구 성암로 성암로184 DMC역 7번 출구
위도:  37.5768239035855
경도: 126.899020186183
인덱스 2287의 데이터의 위도와 경도를 업데이트했습니다.
주소:  마포구 월드컵로 월드컵로 59 기업은행 앞
위도:  37.5543778664217
경도: 126.911268497398
인덱스 2290의 데이터

# 나머지 데이터 `자치구명 +” “ +도로명+” “+ 상세주소`를 카카오 맵 api에 `키워드`로 검색
- 위도, 경도가 없다면
- 주소로 위도, 경도 알아내기
- 카카오 맵 api에 검색



In [21]:
import requests

# Kakao Map API 키
KAKAO_API_KEY = '24a11401d72ce55f0cc05a169bec666e'


# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# 위도와 경도가 없는 데이터를 카카오 맵 API로 검색하여 업데이트
for index, row in df.iterrows():
    위도 = row['위도']
    경도 = row['경도']
    자치구명 = row['자치구명']
    도로명 = row['도로명']
    상세주소 = row['상세주소']

    # 위도와 경도가 없는 경우
    if pd.isnull(위도) or pd.isnull(경도):
        주소 = f'{자치구명} {도로명} {상세주소}'

        # Kakao Map API로 주소 검색
        url = f'https://dapi.kakao.com/v2/local/search/keyword.json?query={상세주소}'
        headers = {'Authorization': f'KakaoAK {KAKAO_API_KEY}'}
        response = requests.get(url, headers=headers)

        # API 응답 확인
        if response.status_code == 200:
            data = response.json()
            if data['documents']:
                # 첫 번째 결과의 위도와 경도를 업데이트
                df.at[index, '위도'] = data['documents'][0]['y']
                df.at[index, '경도'] = data['documents'][0]['x']
                print(f'인덱스 {index}의 데이터의 위도와 경도를 업데이트했습니다.')
                print(f'주소: ', 주소)
                print(f'위도: ', df.at[index, '위도'])
                print(f'경도:',  df.at[index, '경도'])
                updated_count += 1
            else:
                print(f'{주소}의 데이터에 대한 검색 결과가 없습니다.')
        else:
            print(f'인덱스 {index}의 데이터에 대한 API 호출에 실패했습니다.')

# 업데이트된 데이터의 개수 출력
print(f'업데이트된 데이터의 개수: {updated_count}')


스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
인덱스 2883의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 경인로 경인로572
위도:  37.5032690729499
경도: 126.881006458503
인덱스 2884의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 구로중앙로 구로중앙로 68
위도:  37.4958289897032
경도: 126.888839103311
인덱스 2885의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 구로중앙로 구로중앙로28길 65
위도:  37.50038177078585
경도: 126.88883687802908
구로구 구로중앙로 구로중앙로28길 108의 데이터에 대한 검색 결과가 없습니다.
인덱스 2887의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 구로중앙로 구로동514-2
위도:  37.4985242396083
경도: 126.88635623368
인덱스 2888의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 새말로 구로동3-64
위도:  37.5073638319324
경도: 126.892925294001
인덱스 2889의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 공원로 공원로27
위도:  37.49933160529627
경도: 126.89120990191064
인덱스 2890의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 오리로 오리로15길 32
위도:  37.4938815927186
경도: 126.831478611977
인덱스 2891의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 오리로 오리로15길 32
위도:  37.4938815927186
경도: 126.831478611977
인덱스 2892의 데이터의 위도와 경도를 업데이트했습니다.
주소:  구로구 오리로 오리로 1307
위도:  37.4980982768864
경도: 126.82895522097
인덱스 2893의 데이터의 위도와 경도를 업데이트했

# 나머지 데이터 `자치구명 +” “ +도로명+” “+ 상세주소`를 카카오 맵 api에 `주소`로 검색 `(앞, 옆, 뒤 삭제)`
- 위도, 경도가 없다면
- 주소로 위도, 경도 알아내기
- 카카오 맵 api에 검색



In [ ]:
import requests

# Kakao Map API 키
KAKAO_API_KEY = '24a11401d72ce55f0cc05a169bec666e'


# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# 위도와 경도가 없는 데이터를 카카오 맵 API로 검색하여 업데이트
for index, row in df.iterrows():
    위도 = row['위도']
    경도 = row['경도']
    자치구명 = row['자치구명']
    도로명 = row['도로명']
    상세주소 = row['상세주소']

    # 위도와 경도가 없는 경우
    if pd.isnull(위도) or pd.isnull(경도):
        주소 = f'{자치구명} {도로명} {상세주소}'
        주소 = 주소.replace('앞', ' ').replace('옆', ' ').replace('뒤', ' ')

        # Kakao Map API로 주소 검색
        url = f'https://dapi.kakao.com/v2/local/search/address.json?query={상세주소}'
        headers = {'Authorization': f'KakaoAK {KAKAO_API_KEY}'}
        response = requests.get(url, headers=headers)

        # API 응답 확인
        if response.status_code == 200:
            data = response.json()
            if data['documents']:
                # 첫 번째 결과의 위도와 경도를 업데이트
                df.at[index, '위도'] = data['documents'][0]['y']
                df.at[index, '경도'] = data['documents'][0]['x']
                print(f'인덱스 {index}의 데이터의 위도와 경도를 업데이트했습니다.')
                print(f'주소: ', 주소)
                print(f'위도: ', df.at[index, '위도'])
                print(f'경도:',  df.at[index, '경도'])
                updated_count += 1
            else:
                print(f'{주소}의 데이터에 대한 검색 결과가 없습니다.')
        else:
            print(f'인덱스 {index}의 데이터에 대한 API 호출에 실패했습니다.')

# 업데이트된 데이터의 개수 출력
print(f'업데이트된 데이터의 개수: {updated_count}')


# 나머지 데이터 `자치구명 +” “ +도로명+” “+ 상세주소`를 카카오 맵 api에 `키워드`로 검색 `(앞, 옆, 뒤 삭제)`
- 위도, 경도가 없다면
- 주소로 위도, 경도 알아내기
- 카카오 맵 api에 검색



In [ ]:
import requests

# Kakao Map API 키
KAKAO_API_KEY = '24a11401d72ce55f0cc05a169bec666e'


# 업데이트된 데이터의 개수를 저장할 변수 초기화
updated_count = 0

# 위도와 경도가 없는 데이터를 카카오 맵 API로 검색하여 업데이트
for index, row in df.iterrows():
    위도 = row['위도']
    경도 = row['경도']
    자치구명 = row['자치구명']
    도로명 = row['도로명']
    상세주소 = row['상세주소']

    # 위도와 경도가 없는 경우
    if pd.isnull(위도) or pd.isnull(경도):
        주소 = f'{자치구명} {도로명} {상세주소}'
        print(주소)
        주소 = 주소.replace('앞', ' ').replace('옆', ' ').replace('뒤', ' ')
        print(주소)

        # Kakao Map API로 주소 검색
        url = f'https://dapi.kakao.com/v2/local/search/keyword.json?query={상세주소}'
        headers = {'Authorization': f'KakaoAK {KAKAO_API_KEY}'}
        response = requests.get(url, headers=headers)

        # API 응답 확인
        if response.status_code == 200:
            data = response.json()
            if data['documents']:
                # 첫 번째 결과의 위도와 경도를 업데이트
                df.at[index, '위도'] = data['documents'][0]['y']
                df.at[index, '경도'] = data['documents'][0]['x']
                print(f'인덱스 {index}의 데이터의 위도와 경도를 업데이트했습니다.')
                print(f'주소: ', 주소)
                print(f'위도: ', df.at[index, '위도'])
                print(f'경도:',  df.at[index, '경도'])
                updated_count += 1
            else:
                print(f'{주소}의 데이터에 대한 검색 결과가 없습니다.')
        else:
            print(f'인덱스 {index}의 데이터에 대한 API 호출에 실패했습니다.')

# 업데이트된 데이터의 개수 출력
print(f'업데이트된 데이터의 개수: {updated_count}')


# (확인용)위도, 경도 알아낸 데이터 개수 구하기

In [35]:
from datetime import datetime

# 정류장 번호 컬럼 삭제
df.drop(columns=['정류장번호'], inplace=True)

# 현재 날짜 및 시간 생성
current_datetime = datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')

# created_at, modified_at, is_deleted 컬럼 추가 및 값 할당
df['created_at'] = current_datetime
df['modified_at'] = current_datetime
df['is_deleted'] = 0

df

trashcan_id gu_name road_name                         address  \
0               1     종로구       사직로                       경복궁역 4번출구   
1               2     종로구       사직로                       경복궁역 4번출구   
2               3     종로구      자하문로                         자하문로 28   
3               4     종로구      자하문로                         자하문로 28   
4               5     종로구      자하문로                         자하문로 44   
...           ...     ...       ...                             ...   
4894         5371     도봉구       해등로            해등로3길 41 삼환빌라 앞 횡단보도   
4895         5372     도봉구       해등로          해등로 103  창원초등학교 앞 횡단보도   
4896         5373     도봉구      도봉산길         도봉산길 27 도봉고등학교 건너편 횡단보도   
4897         5374     도봉구       도봉로  도봉동 620-23 신도봉사거리 도봉중학교방향 횡단보도   
4898         5375     도봉구       해등로        해등로 32 창1동 서울가든아파트 버스정류장   

     location_type trashcan_type          latitude         longitude  \
0          지하철역 입구            일반  37.5760042456058  126.973261106118   
1          지하철역 입구           재활용  37.5760042456058  126.973261106118   
2          도로(가로)변            일반         37.578567         126.97197   
3          도로(가로)변           재활용         37.578567         126.97197   
4          도로(가로)변            일반         37.579972         126.97155   
...            ...           ...               ...               ...   
4894       도로(가로)변            일반         37.645712        127.042082   
4895       도로(가로)변            일반         37.652644         127.03998   
4896       도로(가로)변            일반         37.686755        127.042578   
4897       도로(가로)변            일반          37.67098        127.043439   
4898         버스정류장            일반         37.501074        126.860358   

                      created_at                 modified_at  is_deleted  
0     2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
1     2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
2     2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
3     2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
4     2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
...                          ...                         ...         ...  
4894  2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
4895  2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
4896  2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
4897  2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  
4898  2023-10-30 04:44:26.601680  2023-10-30 04:44:26.601680           0  

[3076 rows x 11 columns]

In [32]:
# 위도와 경도 열이 비어 있지 않은 데이터의 총 개수 구하기
non_null_coordinates_count = len(df.dropna(subset=['위도', '경도']))

print(f"위도와 경도가 알려진 데이터의 총 개수: {non_null_coordinates_count}개")

위도와 경도가 알려진 데이터의 총 개수: 3076개


# (확인용)위도, 경도를 알아내지 못한 데이터 개수 + 정보


In [ ]:
no_coordinates_count = len(df[(df['위도'].isnull() | df['경도'].isnull())])

print(f"위도와 경도가 없는 데이터의 총 개수: {no_coordinates_count}개")


# 위도와 경도가 NaN인 데이터 출력
missing_coordinates_data = df[ (df['위도'].isnull() | df['경도'].isnull())]

print("위도와 경도가 없는 데이터:")
missing_coordinates_data

# ERD 형태로 csv 재작성
  `ERD` : 공공데이터
- `trashcan_id` : 연번
- `latitude` : 위도
- `longitude` : 경도
- `trashcan_type` : 종류
- `location_type` : 설치지점
- `gu_name` : 자치구명
- `road_name` : 도로명
- `address` : 상세주소
- `created_at` :
- `modified_at` :
- `is_deleted` :


In [ ]:
# 위도와 경도가 없는 데이터 삭제
df = df.dropna(subset=['위도', '경도'])
df


In [34]:
# 컬럼명 변경
df.rename(columns={
    '연번': 'trashcan_id',
    '위도': 'latitude',
    '경도': 'longitude',
    '종류': 'trashcan_type',
    '설치지점': 'location_type',
    '자치구명': 'gu_name',
    '도로명': 'road_name',
    '상세주소': 'address'
}, inplace=True)

# 변경된 데이터프레임 확인
df

trashcan_id gu_name road_name                         address  \
0               1     종로구       사직로                       경복궁역 4번출구   
1               2     종로구       사직로                       경복궁역 4번출구   
2               3     종로구      자하문로                         자하문로 28   
3               4     종로구      자하문로                         자하문로 28   
4               5     종로구      자하문로                         자하문로 44   
...           ...     ...       ...                             ...   
4894         5371     도봉구       해등로            해등로3길 41 삼환빌라 앞 횡단보도   
4895         5372     도봉구       해등로          해등로 103  창원초등학교 앞 횡단보도   
4896         5373     도봉구      도봉산길         도봉산길 27 도봉고등학교 건너편 횡단보도   
4897         5374     도봉구       도봉로  도봉동 620-23 신도봉사거리 도봉중학교방향 횡단보도   
4898         5375     도봉구       해등로        해등로 32 창1동 서울가든아파트 버스정류장   

     location_type trashcan_type  정류장번호          latitude         longitude  
0          지하철역 입구            일반    NaN  37.5760042456058  126.973261106118  
1          지하철역 입구           재활용    NaN  37.5760042456058  126.973261106118  
2          도로(가로)변            일반    NaN         37.578567         126.97197  
3          도로(가로)변           재활용    NaN         37.578567         126.97197  
4          도로(가로)변            일반    NaN         37.579972         126.97155  
...            ...           ...    ...               ...               ...  
4894       도로(가로)변            일반    NaN         37.645712        127.042082  
4895       도로(가로)변            일반    NaN         37.652644         127.03998  
4896       도로(가로)변            일반    NaN         37.686755        127.042578  
4897       도로(가로)변            일반    NaN          37.67098        127.043439  
4898         버스정류장            일반    NaN         37.501074        126.860358  

[3076 rows x 9 columns]

In [49]:
# 모든 컬럼이 다 채워져 있는지 확인
if df.isnull().values.any():
    print("데이터프레임에 빈 값 또는 결측치가 있습니다.")

    # 빈 값이 있는 행 출력
    print("빈 값이 있는 행:")
    print(df[df.isnull().any(axis=1)])
else:
    print("데이터프레임의 모든 컬럼이 다 채워져 있습니다.")


데이터프레임의 모든 컬럼이 다 채워져 있습니다.


# csv 파일로 저장

In [40]:
# df.to_csv("trashcan_data_4.csv",sep=",", index= False)

df.to_csv('drive/MyDrive/Colab Notebooks/trashcan/trashcan_data_15.csv', index=False)